File: src/nl/rubensten/texifyidea/completion/handlers/LatexCommandArgumentInsertHandler.java
package nl.rubensten.texifyidea.completion.handlers;

import com.intellij.codeInsight.completion.InsertHandler;
import com.intellij.codeInsight.completion.InsertionContext;
import com.intellij.codeInsight.lookup.LookupElement;
import com.intellij.openapi.editor.CaretModel;
import com.intellij.openapi.editor.Document;
import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.util.TextRange;
import nl.rubensten.texifyidea.lang.LatexMathCommand;
import nl.rubensten.texifyidea.lang.LatexNoMathCommand;
import nl.rubensten.texifyidea.psi.LatexCommands;

import java.util.List;

/**
 * @author Ruben Schellekens
 */
public class LatexCommandArgumentInsertHandler implements InsertHandler<LookupElement> {

    @Override
    public void handleInsert(InsertionContext insertionContext, LookupElement lookupElement) {
        Object object = lookupElement.getObject();

        if (object instanceof LatexCommands) {
            insertCommands((LatexCommands)object, insertionContext);
        }
        else if (object instanceof LatexMathCommand) {
            insertMathCommand((LatexMathCommand)object, insertionContext);
        }
        else if (object instanceof LatexNoMathCommand) {
            insertNoMathCommand((LatexNoMathCommand)object, insertionContext);
        }
    }

    private void insertCommands(LatexCommands commands, InsertionContext context) {
        List<String> optional = commands.getOptionalParameters();
        if (optional.isEmpty()) {
            return;
        }

        int cmdParameterCount = 0;
        if (!optional.isEmpty()) {
            try {
                cmdParameterCount = Integer.parseInt(optional.get(0));
            }
            catch (NumberFormatException ignore) {
            }
        }

        if (cmdParameterCount > 0) {
            insert(context, "{}");
        }
    }

    private void insertMathCommand(LatexMathCommand mathCommand, InsertionContext context) {
        if (mathCommand.autoInsertRequired()) {
            insert(context, mathCommand.getCommand());
        }
    }

    private void insertNoMathCommand(LatexNoMathCommand noMathCommand, InsertionContext context) {
        if (noMathCommand.autoInsertRequired()) {
            insert(context, noMathCommand.getCommand());
        }
    }

    private void insert(InsertionContext context, String text) {
        Editor editor = context.getEditor();
        Document document = editor.getDocument();
        CaretModel caret = editor.getCaretModel();
        int offset = caret.getOffset();

        // When not followed by {}, insert {}.
        if (offset >= document.getTextLength() - 1 ||
                !document.getText(TextRange.from(offset, 1)).equals("{")) {
            insertSquigglyBracketPair(editor, caret);
        }
        else {
            skipSquigglyBrackets(editor, caret);
        }
    }

    private void skipSquigglyBrackets(Editor editor, CaretModel caret) {
        Document document = editor.getDocument();
        int offset = caret.getOffset();

        int depth = 0;
        for (int i = offset; i < editor.getDocument().getTextLength(); i++) {
            String text = document.getText(TextRange.from(i, 1));

            switch (text) {
                case "{":
                    depth++;
                    break;
                case "}":
                    if (--depth == 0) {
                        caret.moveToOffset(i + 1);
                        return;
                    }
                default:
            }
        }
    }

    private void insertSquigglyBracketPair(Editor editor, CaretModel caret) {
        editor.getDocument().insertString(caret.getOffset(), "{}");
        caret.moveToOffset(caret.getOffset() + 1);
    }
}

File: src/nl/hannahsten/texifyidea/completion/handlers/LatexCommandArgumentInsertHandler.kt
package nl.hannahsten.texifyidea.completion.handlers

import com.intellij.codeInsight.completion.InsertHandler
import com.intellij.codeInsight.completion.InsertionContext
import com.intellij.codeInsight.lookup.LookupElement
import com.intellij.codeInsight.template.TemplateManager
import com.intellij.codeInsight.template.impl.TemplateImpl
import com.intellij.codeInsight.template.impl.TextExpression
import com.intellij.openapi.editor.Editor
import com.intellij.openapi.util.TextRange
import nl.hannahsten.texifyidea.lang.commands.Argument
import nl.hannahsten.texifyidea.lang.commands.RequiredArgument
import nl.hannahsten.texifyidea.psi.LatexCommands
import nl.hannahsten.texifyidea.util.files.psiFile
import nl.hannahsten.texifyidea.util.parser.endOffset
import nl.hannahsten.texifyidea.util.parser.parentOfType

/**
 * @author Hannah Schellekens
 */
class LatexCommandArgumentInsertHandler(val arguments: List<Argument>? = null) : InsertHandler<LookupElement> {

    override fun handleInsert(insertionContext: InsertionContext, lookupElement: LookupElement) {
        insert(insertionContext, lookupElement)
    }

    private fun insert(context: InsertionContext, lookupElement: LookupElement) {
        val editor = context.editor
        val document = editor.document
        val caret = editor.caretModel
        val offset = caret.offset
        // When not followed by { or [ (whichever the first parameter starts with) insert the parameters.
        if (arguments != null && (
                offset >= document.textLength - 1 || document.getText(TextRange.from(offset, 1)) !in setOf("{", "[")
                )
        ) {
            insertParametersLiveTemplate(editor)
        }
        else {
            skipParameters(editor, lookupElement)
        }
    }

    private fun insertParametersLiveTemplate(editor: Editor) {
        // arguments is not null, we checked when calling this function.
        val template = TemplateImpl(
            "",
            arguments!!.mapIndexed { index: Int, argument: Argument ->
                if (argument is RequiredArgument) "{\$__Variable$index\$}" else "[\$__Variable$index\$]"
            }.joinToString(""),
            ""
        )
        repeat(arguments.size) { template.addVariable(TextExpression(""), true) }
        TemplateManager.getInstance(editor.project).startTemplate(editor, template)
    }

    private fun skipParameters(editor: Editor, lookupElement: LookupElement) {
        val document = editor.document
        val file = document.psiFile(editor.project ?: return)
        val caret = editor.caretModel
        val extraSpaces = lookupElement.lookupString.takeLastWhile { it == ' ' }.length
        val psiElement = file?.findElementAt(caret.offset + extraSpaces) ?: return
        if (psiElement.text in setOf("{", "[")) {
            caret.moveToOffset((psiElement.parentOfType(LatexCommands::class)?.endOffset() ?: return) - extraSpaces)
        }
    }
}